secretsmanager/include/secrets_manager.h

```diff
@@ -0,0 +1,73 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include <string>
+#include <memory>
+#include <set>
+#include <functional>
+
+#include "Logger.hpp"
+
+#include "SecretRetriever.hpp"
+#include "secrets_manager_impl.h"
+#include "secrets_manager_plain.h"
+#include "secrets_manager_enc.h"
+namespace concord::secretsmanager {
+
+// SecretsManager is a singleton which transparently handles encryption and decryption of files on the local filesystem.
+// It wraps twp implementations - SecretsManagerPlain and SecretsManagerEnc. The former is dummy implementation for the
+// cases where encryption is disabled. It only reads a text file and returns it as a buffer. The latter handles the
+// actual implementation. Before using the singleton, init() should be called with proper configuration. There are two
+// configuration parameters:
+// - secrets_manager_mode (bool) - if encryption is enabled or disabled
+// - secrets_url - URL to the Secrets API.
+//
+// About the Secrets API URL: Might be file:// or http(s) endpoint. The former is mainly used for development tests, the
+// latter is intended for actual use. Check tests/secrets_api_mock for a sample.
+class SecretsManager {
+  inline static std::unique_ptr<SecretsManager> instance;
+
+  std::unique_ptr<ISecretsManagerImpl> impl_;
+  SecretsManager(std::unique_ptr<ISecretsManagerImpl> impl) : impl_{std::move(impl)} {}
+
+ public:
+  static SecretsManager& get() {
+    if (instance)
+      return *instance;
+    else
+      throw std::runtime_error("SecretsManager is not initialised");
```

Comment:
Here me and @KTeofanidis  made an unfair assumption (unfair, because it is in our heads only). The SecretsManager purpose is to encrypt configuration. The process can't do much until its configuration is decrypted, so init() will be  called before get().

But I agree that we might introduce a nasty issues in the future in this code.

Considering all the feedback - we'll remove the singleton. I'll update the PR later today.